vllm/config/utils.py

```
@@ -27,3 +38,103 @@ def config(cls: ConfigT) -> ConfigT:
     """
     return cls
+
+
+def normalize_value(x):
+    """Return a stable, JSON-serializable canonical form for hashing.
+
+    Order: primitives, special types (Enum, callable, torch.dtype, Path), then
+    generic containers (Mapping/Set/Sequence) with recursion.
+    """
+    # Fast path
+    if x is None or isinstance(x, (bool, int, float, str)):
+        return x
+
+    # Enums by value
+    if isinstance(x, enum.Enum):
+        return x.value
+
+    # Callable by qualified name
+    try:
+        if callable(x):
+            module = getattr(x, "__module__", "")
+            qual = getattr(x, "__qualname__", repr(x))
+            return f"{module}.{qual}" if module else qual
+    except Exception:
+        pass
+
+    # Torch dtype without hard dependency
+    try:
+        import torch
+        if isinstance(x, torch.dtype):
+            return str(x)
+    except Exception:
+        pass
+
+    # Bytes
+    if isinstance(x, (bytes, bytearray)):
+        return x.hex()
+
+    # Paths (canonicalize)
+    if isinstance(x, pathlib.Path):
+        try:
+            return str(x.expanduser().resolve())
+        except Exception:
+            return str(x)
+
+    # Containers (generic)
+    if isinstance(x, Mapping):
+        return tuple(sorted(
+            (str(k), normalize_value(v)) for k, v in x.items()))
+    if isinstance(x, Set):
+        return tuple(sorted(repr(normalize_value(v)) for v in x))
+    if isinstance(x, Sequence) and not isinstance(x, (str, bytes, bytearray)):
+        return tuple(normalize_value(v) for v in x)
+
+    # Nested configs which provide a uuid() method
+    if hasattr(x, "uuid") and callable(x.uuid):
+        return x.uuid()
+
+    # PretrainedConfig
+    if hasattr(x, "to_json_string") and callable(x.to_json_string):
+        return x.to_json_string()
+
+    # Unsupported type
+    with suppress(Exception):
+        logger.debug("normalize_value: unsupported type '%s'",
+                     type(x).__name__)
```

Why suppress Exception?